woocommerce/woocommerce-ios
# REST API: Integrate authentication link request

Part of #8918 


## Description

This PR continues the WPCom login flow of the Jetpack setup feature with the following changes:
- `AppDelegate`:
  - Check if the user is authenticated without WPCom and create a coordinator for Jetpack setup to handle the URL if possible.
- `JetpackSetupCoordinator`:
  - Replaced `navigationController` with `rootViewController` of type `UIViewController`. The reason is that we don't push any view to the existing navigation stack, only present new modals; and `UIViewController` makes it easier to initialize the coordinator from `AppDelegate` for authentication URL handling.
  - Added new method `startAuthentication` to either request a magic link for passwordless accounts or present the password login screen. Both UI for these screens will be added in a future PR.
  - Added new methods `handleAuthenticationUrl` to handle magic link and `startJetpackSetupFlow` to check the Jetpack connection status and proceed with the Jetpack setup flow.
  - Error handling for network requests.

## Testing instructions

1. Testing magic link handling:
Prerequisite: Make sure that you have access to a site with WooCommerce and no Jetpack.
- Log out of the app or skip onboarding if needed.
- Tap "Enter your site address" on the prologue screen and proceed with the address of your test site.
- Log in with the credentials of an admin.
- When the login completes, tap the Jetback benefit banner at the bottom of the Dashboard screen.
- Tap Log In on the benefit modal.
- Enter the email address of a password-less WPCom account on the presented WPCom email login screen.
- Tap the Log In button in the received email (or paste the link of that button in a browser on your device).
- Notice that Woo is opened and a loading screen is displayed briefly before dismissed.
- Notice in Xcode console `✅ Ready for Jetpack setup - connection only: true/false` - make sure that the correct state of the Jetpack plugin is logged.

2. Testing error handling:
- Suppose you're still logged in to the same site and account in the above test case, follow the steps above to show the benefit modal.
- Disable your network connection and tap Log In on the benefit modal. Notice that an error alert is displayed with the message "Error checking the Jetpack connection on your site".
- Enable the network connection again and tap Log In.
- Enter your WPCom email address on the presented email login screen.
- Disable your network connection again and tap Install Jetpack. Notice that an error alert is displayed with the message "Error checking the WordPress.com account associated with this email. Please try again."
- Please feel free to test the magic link request as well; it's probably easier to enable network connection and mock the response for `auth/send-login-email` endpoint.

## Screenshots

N/A

---
